### DISCUSSION TITLE
React draggable layer console error; `deckRef.current.setProps is not a function`

### DISCUSSION THREAD

**Discussion by jessekrubin** [visgl/deck.gl#6689] [Q&A]
I have figured out how to make a layer draggable with using the DeckGL react lib (`@deck.gl/react`) but I am unable to call the `setProps` method of my DeckGL ref without getting the error below in the chrome/firefox console. Any thoughts or help would be greatly appreciated. I would like it if I did not have to not monkey-patch the deck lib or write my own react wrapper.

Bug repro: [jessekrubin/deckgl-drag-setprops-typeerror](https://github.com/jessekrubin/deckgl-drag-setprops-typeerror)

When I use a deckRef with react and deckgl in a 'draggable' layer (which totally works), and try to call setProps, I get the error below (in the chrome and firefox console(s)); the error cannot be try/catch-ed and doing a try/catch on the setProps call makes the layer no-longer draggable.

ASIDE FROM THE ERROR IN THE BROWSER CONSOLE EVERYTHNG WORKS FINE!?!?!?!?!?!?

ERROR:

```
App.js:50 Uncaught TypeError: deckRef.current.setProps is not a function
    at IconLayer.onDragStart (App.js:50:1)
    at Deck._onEvent (deck.js:817:1)
    at EventRegistrar._emit (event-registrar.js:121:1)
    at EventRegistrar.handleEvent (event-registrar.js:96:1)
    at Manager.emit (hammer.js:2519:1)
    at emit (hammer.js:1527:1)
    at PanRecognizer.emit (hammer.js:1532:1)
    at PanRecognizer.emit (hammer.js:1817:1)
    at PanRecognizer.tryEmit (hammer.js:1555:1)
    at PanRecognizer.recognize (hammer.js:1602:1)
```

**Comment by Pessimistress** [comment]
https://deck.gl/docs/api-reference/react/deckgl#methods

> For example, to rerender your component, you can pass updated props to `DeckGL` component directly, while you should call setProps with new props in `Deck`.

```jsx
function App() {
  const [controller, setController] = useState(true);

  const iconlayer = new IconLayer({
    ...
    onDragStart: (obj) => {
      setController({dragPan: false});
    },
    onDragEnd: () => {
      setController({dragPan: true});
    }
  });

  return (
    <DeckGL
      initialViewState={INITIAL_VIEW_STATE}
      controller={controller}
      layers={[iconlayer]}
    />
  );
}
```

**Reply by jessekrubin** [reply]
@Pessimistress Thank you for your response! very helpful. I was pretty sure I had tried that last night and it did not work...  most likely I was fried/tired.

I also was able to get it to work with the following and I feel like it might be a bit snappier feeling:

```jsx
function App() {
  const [controller, setController] = useState(true);

  const iconlayer = new IconLayer({
    ...
    onDragStart: (obj, event) => {
      if (obj.layer !== null){  // if is the layer I want (not null or something else here)...
        event.stopImmediatePropagation ()
      }
    },
  });

  return (
    <DeckGL
      initialViewState={INITIAL_VIEW_STATE}
      controller={controller}
      layers={[iconlayer]}
    />
  );
}
```

@Pessimistress Do you have any do you think one should be preferred over the other? also (possible noob question) do you have any idea why I was getting that typeError from calling `setProps`? and why was try/catch-ing it didn't just silence the error?

**Reply by Pessimistress** [reply]
There is no `setProps` method exposed on the React ref. Because you are not supposed to call it, period. If you choose to use React you need to update the props the reactive way. Calling any imperative API will break the binding between React props and the underlying state.
